### PR TITLE
UM API to write into a ring buffer map.

### LINKED ISSUE #3572: New eBPF API to insert events into ring buffer eBPF map.
A common use case of an eBPF ring buffer map to use as a queue of events generated by eBPF programs that can be read in-order by user space applications. The API request is to let user space applications to write events into the ring buffer map, such that events by eBPF programs in kernel and user space applications are interleaved in the order they were generated, that can be consumed by another user space application

### PR DESCRIPTION
**Pull request by shankarseal**
## Description

Fixes #3572.

Add a new API in eBPF API.dll to write into a ring buffer map from user mode. This API issues an IOCTL which calls `bpf_ring_buffer_output` function to write user specified data into the ring buffer.

## Testing

Updated unit and api tests to write messages into ring buffer interleaved with events that invoke eBPF programs (which in turn write into the same ring buffer).

## Documentation

No documentation impact.

## Installation

No installation impact.

### CODE REVIEW

File: include/ebpf_api.h [hunk 1]
```diff
@@ -564,6 +564,20 @@ extern "C"
     _Must_inspect_result_ ebpf_result_t
     ebpf_program_test_run(fd_t program_fd, _Inout_ ebpf_test_run_options_t* options) EBPF_NO_EXCEPT;
 
+    /**
+     * @brief Write data into the ring buffer map.
```
Comment (dthaler) [inline include/ebpf_api.h:568]:
```suggestion
     * @brief Write data into a ring buffer map.
```
Also I think "ring buffer map" should be expanded to say what type of ring buffer map, since this shouldn't be permitted in all ring buffer maps.

File: include/ebpf_api.h [hunk 2]
```diff
@@ -564,6 +564,20 @@ extern "C"
...
+     *
+     * @param ring_buffer_map_fd ring buffer map file descriptor.
+     * @param data Pointer to data to be written.
+     * @param data_length Length of data to be written.
```
Comment (dthaler) [inline include/ebpf_api.h:572]: Missing `[in]`

File: tests/libs/common/common_tests.cpp [hunk 3]
```diff
@@ -189,17 +189,30 @@ void
 ring_buffer_api_test_helper(
```
Comment (dthaler) [inline tests/libs/common/common_tests.cpp:189]: Also add a test case verifying that the API _fails_ if the caller passes an fd that is not an appropriate ring buffer map.

File: libs/execution_context/ebpf_core.c [hunk 4]
```diff
@@ -2073,6 +2073,25 @@ _ebpf_core_protocol_ring_buffer_map_async_query(
...
+static ebpf_result_t
+_ebpf_core_protocol_ring_buffer_map_write_data(_In_ const ebpf_operation_ring_buffer_map_write_data_request_t* request)
+{
+    ebpf_map_t* map = NULL;
+    ebpf_result_t result =
+        EBPF_OBJECT_REFERENCE_BY_HANDLE(request->map_handle, EBPF_OBJECT_MAP, (ebpf_core_object_t**)&map);
+    if (result != EBPF_SUCCESS) {
+        goto Exit;
+    }
+    if (ebpf_map_get_definition(map)->type != BPF_MAP_TYPE_RINGBUF) {
```
Comment (dthaler) [inline libs/execution_context/ebpf_core.c:2085]: I don't think this is sufficient since it would work with any ring buffer map, which I don't think is appropriate.
It should also check (say) a flag on the ring buffer map, or use a separate map type.

File: ebpfapi/Source.def [hunk 5]
```diff
@@ -140,3 +140,4 @@ EXPORTS
     libbpf_strerror
     ring_buffer__new
     ring_buffer__free
+    ebpf_ring_buffer_map_write
```
Comment (dthaler) [inline ebpfapi/Source.def:143]: Move up in alphabetical order

File: include/ebpf_api.h (hunk 1 above)
Comment (dthaler) [reply include/ebpf_api.h:568]: Resolving this per meeting discussion, should use flags.

File: libs/execution_context/ebpf_core.c (hunk 4 above)
Comment (dthaler) [reply libs/execution_context/ebpf_core.c:2085]: Resolving per meeting discussion.  Will file separate issue.

File: libs/execution_context/ebpf_protocol.h [hunk 6]
```diff
@@ -390,6 +391,14 @@ typedef struct _ebpf_operation_ring_buffer_map_async_query_reply
     ebpf_ring_buffer_map_async_query_result_t async_query_result;
 } ebpf_operation_ring_buffer_map_async_query_reply_t;
 
+typedef struct _ebpf_operation_ring_buffer_map_write_data_request
+{
+    struct _ebpf_operation_header header;
+    ebpf_handle_t map_handle;
+    size_t data_length;
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_protocol.h:398]: Why is this needed? Header already contains total request length, so this is redundant.

File: libs/execution_context/ebpf_core.c [hunk 7]
```diff
@@ -2073,6 +2083,30 @@ _ebpf_core_protocol_ring_buffer_map_async_query(
...
+        result = EBPF_INVALID_ARGUMENT;
+        EBPF_LOG_MESSAGE_ERROR(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_CORE,
+            "write data operation called on a map that is not of the ring buffer type.",
+            result);
+        goto Exit;
+    }
+    result = ebpf_ring_buffer_map_output(map, (uint8_t*)request->data, request->data_length);
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_core.c:2104]: Where is the validation that the length of the data is valid?

**Review by Alan-Jowett — CHANGES_REQUESTED** [review]
Need to validate the request.

File: libs/execution_context/ebpf_core.c [hunk 8]
```diff
@@ -2073,6 +2083,32 @@ _ebpf_core_protocol_ring_buffer_map_async_query(
...
+    if (ebpf_map_get_definition(map)->type != BPF_MAP_TYPE_RINGBUF) {
+        result = EBPF_INVALID_ARGUMENT;
+        EBPF_LOG_MESSAGE_ERROR(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_CORE,
+            "write data operation called on a map that is not of the ring buffer type.",
+            result);
+        goto Exit;
+    }
+    data_length = request->header.length - EBPF_OFFSET_OF(ebpf_operation_ring_buffer_map_write_data_request_t, data);
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_core.c:2105]: Switch to using safe integer arithmetic. I.e. ebpf_safe_size_t_subtract or similar.

**Review by Alan-Jowett — APPROVED** [review]

File: libs/api/ebpf_api.cpp [hunk 9]
```diff
@@ -4426,6 +4445,62 @@ ebpf_ring_buffer_map_subscribe(
...
+        request->header.length = static_cast<uint16_t>(request_buffer.size());
+        request->header.id = ebpf_operation_id_t::EBPF_OPERATION_RING_BUFFER_MAP_WRITE_DATA;
+        request->map_handle = (uint64_t)map_handle;
+        std::copy((uint8_t*)data, (uint8_t*)data + data_length, request->data);
+
+        result = win32_error_code_to_ebpf_result(invoke_ioctl(request_buffer));
+    } catch (const std::bad_alloc&) {
+        EBPF_RETURN_RESULT(EBPF_NO_MEMORY);
+    }
+    EBPF_RETURN_RESULT(result);
```
Comment (saxena-anurag) [inline libs/api/ebpf_api.cpp:4500]: nit: This API will typically be in hot path. Should we skip tracing in case of success? Same for `EBPF_LOG_ENTRY` at the start of this function.

**Review by saxena-anurag — APPROVED** [review]

File: libs/api/ebpf_api.cpp (hunk 9 above)
Comment (shankarseal) [reply libs/api/ebpf_api.cpp:4500]: `EBPF_RETURN_RESULT` calls `EBPF_LOG_FUNCTION_SUCCESS` which is invoked only at verbose level. We could change the keyword for `EBPF_LOG_FUNCTION_SUCCESS` so that we dont accidentally turn it on.
